README
Simulated annealing (SA)
Copyright (C) 2020  Dr. Juan Paulo Sánchez Hernández, and Dr. Juan Frausto Solis
Copyright (C) 2005 Frank Eisenmenger, U.H.E. Hansmann, Shura Hayryan, Chin-Ku Hu
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
USA.

README
The following lines are from the SMMP README file:

The SMMP  package is designed for molecular simulation of linear 
proteins within the standard geometry model.  The package contains 
various modern Monte Carlo algorithms and energy minimization 
routines.  The energy of the protein can be calculated by exploiting 
one of three force fields: ECEPP/2, ECEPP/3 or FLEX. Two subroutines 
are included for approximating  protein-solvent interaction by means 
of calculating the solvent accessible area of atomic groups (either
analytically or in an approximate way).  More information on SMMP can 
be found in the manual (file `manual.ps'), our papers (F. Eisenmenger, 
U.H.E. Hansmann, S. Hayryan and C.K. Hu, [SMMP] A Modern Package for 
Protein Simulations, Comp.Phys. Comm.  (2001), 138 (2001) 192-212;
[SMMP 1.1] - An Enhanced Open-Source Software Package, submitted), and  
on the SMMP homepage: http://www.smmp05.net

SMMP is offered as open code.  We encourage users to re-write code, 
or add their own modules, whenever they see the need. For terms of 
use of SMMP, please see the manual (manual.pdf) and the license
agreement (license.txt). By viewing or using any part of SMMP you
agree with the terms of the license. Any suggestions for improvement 
of the code or reports on bugs are welcome.
  
Please send your remarks to:
-  Frank Eisenmenger:    eisenmenger@fmp-berlin.de
-  Ulrich H.E. Hansmann: hansmann@mtu.edu 
-  Shura Hayryan:        shura@phys.sinica.edu.tw
-  Chin-Kun Hu:          huck@phys.sinica.edu.tw
-  Jan H. Meinke         j.meinke@fz-juelich.de



2) INSTALLATION:
After  uncompressing and un-tar-ing the SMMP package into a separate
directory, the following files should be in that directory:

- README           This README-file

- SMMP             A sub-directory containing:
  - lib.sh2        Library file with ECEPP/2 parameters
  - lib.sh3        Library file with ECEPP/3 parameters
  - lib.flex       Library file with FLEX parameters
  - charges        File with charges (needed only for FLEX)
  - tes.dat        File with tesselation points needed for approximating
                   the solvent accessible surface area of atoms
- doc              A sub-directory containing:
  - manual.pdf     The manual in PDF-format describing the details
                   of various subroutines and installation of SMMP
  - license.txt    Ascii-file with the license agreement
  - angle_defs.svg  Scalable vector graphic of definition of angles
  - atom_numbering.svg  Scalable vector graphic of numbering of atoms
  - dihedral_defs.svg  Scalable vector graphic of definition of dihedrals
  - Makefile       Makefile to build documentation
  - manual.lyx     Original source of manual
  - manual.tex     Lyx file converted to tex suitable for pdflatex
  
- Makefile         Produces the executable file

- INCL.H           Defines global parameters and common-blocks
- INCP.H           Defines parameters and common-blocks necessary for reading
                   PDB-structures
- incl_lund.h      Defines parameters and common blocks used with the Lund
                   force field and BGS
                   
- init_molecule.f  Wrapping function that constructs the start configuration
                   of a molecule
- init_energy.f    Wrapping function that initializes energy parameter
- redseq.f         Reads sequence of amino acids (e.g. enkefa.seq)
- redvar.f         Reads initial configuration of molecule (e.g. enkefa.var)
- getmol.f         Assembles data from libraries
- bldmol.f         Builds up the protein atomic coordinates
- addend.f         Adds end-groups
- setmvs.f         Determines sets of moving atoms for given variables
- mklist.f         Compiles lists of interaction partners
- setvar.f         Resets variables and rebuilds the molecule
- dihedr.f         Returns dihedral angles and valence angles 
- difang.f         Calculates the difference and the sum of two angles
- nursvr.f         Calculates the residue index of a dihedral angle or atom
- redstr.f         String input routines
- pdbread.f        Reads amino acid sequence and rotein atomic coordinates from 
                   a PDB-file, calculates dihedral angles from PDB coordinates,
                   and builds index field that relates PDB-atoms to SMMP atoms

- energy.f         Wrapping function that returns the energy of a current 
                   protein configuration
- enyflx.f         Calculates internal energy of molecule  with FLEX dataset
- enyshe.f         Calculates internal energy of molecule  with ECEPP datasets
- enyshe_p.f       Parallel version of enyshe.f
- enylun.f         Calculates energy of molecule using the Lund force field
- enysol.f         Calculates solvation energy of molecule using solvent
                   accessible area method (fast, but approximate calculation)
- enysol_p.f       Parallel version of enysol.f
- esolan.f         Calculates solvation energy of molecule using solvent
                   accessible area method (analytic, but slow calculation) 
- eninteract.f     Calculates interaction term for multi-molecule simulations  
                   based on ECEPP.
- enyreg.f         Calculates a constraint energy needed for regularizing 
                   PDB-structures
- eyabgs.f         Calculates correction term introduced by R. A. Abagyan 
                   et al.

- gradient.f       Wrapping function that returns the energy gradient vs.
                   dihedral angles for a protein configuration
- opeflx.f         Calculates internal energy and partial derivatives vs.
                   dihedral angles for FLEX dataset
- opeshe.f         Calculates internal energy and partial derivatives vs.
                   dihedral angles for ECEPP datasets
- opesol.f         Calculates  analytically the partial derivatives vs.  
                   dihedral angles  of the solvation energy
- opereg.f         Calculates the partial derivatives vs. diheadral angles
                   of the constraint energy term during  regularization


- main.f           Main program
- regul.f          Regularization of PDB-structure into SMMP geometry
- anneal.f         For simulated annealing run
- canon.f          For canonical Monte Carlo run
- minim.f          For minimization of protein potential energy
- mulcan_par.f     Calculates multicanonical weights
- mulcan_sim.f     For multicanonical simulation run
- mulcan_par_mod.f90 Combines the previous two files in a Fortran module.
- partem_s.f       For parallel tempering run on a single-processor machine
- partem_p.f       For parallel tempering run on a multiple-processor machine
- main_p.f         Replaces `main.f' on multiple-processor machine in
                   parallel tempering runs.
- main_bgl_p.f     Version of main_p.f optimized for use with IBM BlueGene/L

- metropolis.f     Performs Metropolis updates
- bgs.f            Biased Gaussian step.
- minqsn.f         Minimization by quasi-Newton method using BFGS-formula
- mincjg.f         Minimization by conjugate gradient method
- outvar.f         Output of the current conformation (dihedral angles) 
- contacts.f       Calculates van der Waals contacts
- cnteny.f         Calculates atomic contact energy and prints bad contacts
- hbond.f          Calculates number of hydrogen bonds in a configuration
- helix.f          Measures the number of residues which are part of helix or
                   beta-sheet
- outpdb.f         Output of current configuartion in PDB-format
- rgyr.f           Measures the radius-of-gyration and end-to-end distance in
                   molecule
- zimmer.f         Expresses given configuration in Zimmerman code
- rmsdfun.f        Calculates root-mean-square deviation between current
                   SMMP configuration and a reference structure.

- twister.f        Mersenne-Twister random number generator
- utilties.f       Somer helper function for simulations on multiple processors.

- smmp.pyf         Interface of the Python bindings

- universe.py      Python package that provides access to global properties 
                   of the system.
- protein.py       Python package representing atoms, amino acids, and proteins
- algorithms.py    Some basic algorithms using the Python bindings.
- main.py          Main program in Python

- rmexclpoint.py   Utility script to build the Python binding
- restoreexclpoint.py  Utility script to build the Python binding

- temperatures     Sample temperature file with 32 temperatures.

- EXAMPLES         A sub-directory containing:
  - enkefa.seq     Example sequence file
  - enkefa.var     Example configuration file
  - enkefa.ann     Example configuration filea for simulated annealing run
  - enkefa.ref     Example contact matrix file
  - abeta.seq      Sequence file for Abeta_16-22
  - abeta.var      Configuration file with global coordinates 
  - abeta.ref      Dummy contact matrix
  - 1bdd.pdb       PDB file of protein A for regularization.
  - 1bdd.ref       Contact matrix of protein A
  - 1bdd.seq       Sequence file of protein A
  - 1bdd.var       Sampe configuration of protein A
  - temperatures   Example temperature file for parallel_tempering_s
  - temperatures_abeta Example temperature file for parallel_tempering
  - smmp.cmd       Example shell-script to run smmp
  
  - Makefile       Makefile for building the examples.
  
  - annealing.f    Example for running simulated annealing
  - minimization.f Example for a minimization
  - multicanonical.f  Example for calculating parameters for a multi canonical
                   simulation.
  - parallel_tempering_p.f  Example for parallel tempering on a cluster
  - parallel_tempering_s.f  Example for parallel tempering on a single node
  - partem_p.f     Parallel tempering routine used with parallel_tempering_p.f
                   The output is different from the default implementation.
  - regularization.f Example for regularizing protein A (1bdd)
  
  - Python         A sub-directory containing some examples that use the Python
                   bindings.

  - annealing.py   Python version of the annealing example
  - minimization.py Python version of the minimization example  
  - muca.py        Python version for calculating the parameters for a 
                   multi-canonical simulation. 
  - regularization.py Python version of regularization example
  - gui_example.py Example for building a graphical user interface for SMMP
  - best.pml       A PyMol script for rendering best.pdb 

  - scripts        A sub-directory containing some useful scripts:
    - README       Short description of the scripts
    - atomprops.py Lists the properties of all atoms in a protein
    - var2pdb.py   Takes a sequence and a var file and builds a PDB from it.

The whole SMMP package is  written in standard FORTRAN language. 
We have been exploiting it under pgif, ifort, gfortran, and xlf. 
It should be possible to compile the code with any contemporary fortran 
compilers.  There are no machine dependent routines included in SMMP.  
Common blocks and limiting parameters are gathered in special files 
``INCL.H''  and ``INCP.H'' which are attached to the modules through 
an `include' statement.  In order to install SMMP the user needs to edit the
`Makefile' and specify the compiler and compiler options which he will use. 
Executing the `make' command will finish installation of SMMP. For compiling
the parallel version use `make parallel'. The command `make doc' build the
documentation and `make pybind' build the Python bindings if f2py is installed.



3) HOW TO RUN SMMP:
SMMP does not include an interpretor of user defined commands. 
The  preparation of a simulation must be done in the   'main' module 
After changes of SMMP has to be re-compiled.  

Alternatively, you can use the Python bindings, which allow for interactive
simulations.

The residues which can be used with each parameter set are described in
files `lib.sh2', 'lib.sh3' and 'lib.flx', respectively. The file 'charges' 
is needed for N- and C-terminal residues with FLEX parameters. The 
directory with these 3 files should be given in string 'libdir', which 
is assigned in module 'main' or 'pmain'. 

SMMP  requires as input a file that specifies the sequence of residues
in a protein. This sequence can be read either from a PDB-file (the
standard format in which protein structures are deposited in the
Protein Data Bank) or from a special sequence file. If the sequence
is read from a sequence file its first line must start with a '#' and 
may (or may not) contain the name for the molecule. The residues in the
following lines should be named as in the libraries (not case-
sensitive). Residue names should be separated by at least one space.
An example file ("enkefa.seq") is provided in the sub-directory EXAMPLES.

The initial values for internal variables, i.e. dihedral angles for single
bonds, can be calculated either from the atomic coordinates of the PDB-file,
or (if the sequence is read from a sequence file) may be provided in a 
SECOND INPUT file.  If this file is not given (or the name of a non-existing 
file is entered), all variables retain their values given in the libraries.
The example file ("enkefa.var") which is provided in the subdirectory
EXAMPLES demonstrates the syntax that has to be used:

                   residue(s) : variable(s) : value

In the first field  the RESIDUE is selected through an {\it INTEGER} 
number which marks the position of that RESIDUE in the amino acid sequence. 
The second field lists a string with the name of the VARIABLE, i.e.  names 
the specific dihedral angle. The last field lists the value (a {\it REAL} 
number) for the VARIABLE and is mandatory.  Missing fields are interpreted 
as 'for all'.  Spaces are not significant and are ignored. Empty lines or
or lines containing '#' are ignored. A line containing '&' assigns FIXED 
variable(s), i.e. they will be set to the given value, but will NOT be 
varied during subsequent changes of the protein configuration.



The following steps summarize how to run SMMP:

-  Assign to the character variable  'libdir' the path to the directory 
   containing the standard amino acid residue libraries and the file 'charges'.

-  Select the force field and solvation model by setting the four 
   'sh2', 'epsd' and  'itysol', 'ientyp' to their appropriate values:
   * ientyp      : 0  => ECEPP2 or ECEPP3 depending on the value of sh2
                   1  => FLEX 
                   2  => Lund force field
                   3  => ECEPP with Abagyan corrections
   * sh2 =.TRUE. : ECEPP/2 potential, sh2=.FALSE.:  ECEPP/3 
                   (Note that the variable  `flex' has to be set to .FALSE.)
   * epsd=.TRUE. :  Distant dependent  epsilon(r)
     epsd=.FALSE.: epsilon=2
   * itysol = 0  : Gas phase;  
     itysol > 0  : approximation of protein-solvent interactions by means 
                   of a solvent accessible surface area approach with
                   stochastic estimation of the accessible area.
     itysol < 0  : same as above, but the accessible area is calculated
                   analytically (much SLOWER than itysol > 0).

-  Choose a  N-terminal and C-terminal group by setting 'grpn' and 'grpc'
   to approbriate values.

-  Choose how the initial input is read in:
   * iabin = 0   : read from PDB-file
   * iabin = 1   : read from sequence (and configuration) file

-  Enter the names of the corresponding file(s). In the example 
   version of the  'main' module this is done through interactive dialog 
   but the user can easily just assign the corresponding names to the 
   character variables  'seqfil' and  'varfil' in the subroutine 
   'init_molecule'.  

-  At this point the program is ready for calling the simulation 
   subroutines.  In the provided version the energy minimization 
   subroutine is called through  'call minim'. A detailed description
   of this and other simulation subroutines can be found in the
   manual (file manual.ps). Normally the  simulation subroutines 
   write data in output files, but one can also put output routines 
   such as `outpdb' in 'main'.  The minimal output (written into 
   standard output) is the name of the sequence file (extension .seq), 
   name of configuration file (extension .var), and for each residue 
   a list of dihedral angles together with their initial values.

-  For parallel tempering jobs on on a multiprocessor system one has
   to replace 'main' by 'main_p'. The above protocol still applies.


4) LIMITATIONS:
All parameters which limit the usage of SMMP are stored in the 
file ``INCL.H''.  The most important ones are listed below. 

mxml=10          max. number of molecules
mxrs=500        max. total number of residues
mxat=10000       max. total number of atoms
mxbd=3          max. number of bonds to following atoms
mxvr=mxrs*5     max. number of local variables
mxms=mxvr*3     max. total number of moving sets
mxvw=mxat*4     max. number of vwd domains
mx14=mxat*4     max. number of '1-4' partners
mxath = 100     max. number of atoms in help-arrays
mxvrh=mxath     max. number of variables in help
mxtyat=18       max. number of energetic atom-types
mxhbdo=4        max. types of Hydrogens as donors in HB
mxhbac=6        max. types of atoms as acceptors in HB
mxtyto=19       max. number of types of torsional potentials
nrsty=35        max. number of residue types
mxtysol=9       the number of solvation parameters sets

Note also the following restrictions in the current version of SMMP:
- A single amino acid residue can not be simulated with FLEX potential.
- A protein must not start with a prolin residue.


5) EXAMPLE:
Proper installation of SMMP can be tested by running the following 
example. After compilation of the program package (with `make' 
command using the provided `Makefile') and runing SMMP by typing 
./smmp, SMMP will minimize the ECEPP/3 energy of the Met-enkephalin 
configuration in `EXAMPLES/enkefa.var'. 
-----------------------------------------------------------------------
|NOTE: If the program doesn't start but only shows the error message  |
|"Killed", you probably don't have enough memory available. You can   |
|reduce the memory requirement by setting lower limits for mxrs       |
|(line 10) and mxat (line 11) in the file INCL.H. Setting mxrs=10 and |
|mxat=1300 will still run all the examples.                           |
-----------------------------------------------------------------------
Running the program leads to the  following output:  

------------------------------------------------------------ enkefa.out

 file with SEQUENCE:
 ./EXAMPLES/enkefa.seq

 file with VARIABLES: ./EXAMPLES/enkefa.var

 redvar> Met-Enkephalin: residue    1 Tyr : x1  set   -172.590
 redvar> Met-Enkephalin: residue    1 Tyr : x2  set     78.710
 redvar> Met-Enkephalin: residue    1 Tyr : x6  set   -165.880
 redvar> Met-Enkephalin: residue    1 Tyr : phi set    -86.240
 redvar> Met-Enkephalin: residue    2 Gly : psi set    156.180
 redvar> Met-Enkephalin: residue    2 Gly : omg set   -180.000
 redvar> Met-Enkephalin: residue    2 Gly : phi set   -154.530
 redvar> Met-Enkephalin: residue    3 Gly : psi set     83.640
 redvar> Met-Enkephalin: residue    3 Gly : omg set    180.000
 redvar> Met-Enkephalin: residue    3 Gly : phi set     83.660
 redvar> Met-Enkephalin: residue    4 Phe : psi set    -73.860
 redvar> Met-Enkephalin: residue    4 Phe : omg set   -180.000
 redvar> Met-Enkephalin: residue    4 Phe : x1  set     58.790
 redvar> Met-Enkephalin: residue    4 Phe : x2  set     94.600
 redvar> Met-Enkephalin: residue    4 Phe : phi set   -137.040
 redvar> Met-Enkephalin: residue    5 Met : psi set     19.330
 redvar> Met-Enkephalin: residue    5 Met : omg set   -180.000
 redvar> Met-Enkephalin: residue    5 Met : x1  set     52.760
 redvar> Met-Enkephalin: residue    5 Met : x2  set    175.280
 redvar> Met-Enkephalin: residue    5 Met : x3  set   -179.830
 redvar> Met-Enkephalin: residue    5 Met : x4  set    -58.570
 redvar> Met-Enkephalin: residue    5 Met : phi set   -163.630
 redvar> Met-Enkephalin: residue    5 Met : pst set    160.450
 redvar> Met-Enkephalin: residue    5 Met : omt set    180.000
  

 Energy BEFORE minimization:

 Total:  0.10354E+04
   Coulomb:  0.1991E+02 Lennard-Jones:  0.1142E+03 HB:  0.9008E+03
   Variables:  0.4511E+00  Solvatation:  0.0000E+00

 Step     1: energy  0.489976E+05  ( 0.867243E+13 )
 Step     2: energy  0.108241E+03  ( 0.964364E+07 )
 Step     3: energy -0.632755E+00  ( 0.498256E+04 )
 Step     4: energy -0.739539E+01  ( 0.109979E+05 )
 Step     5: energy  0.517957E+03  ( 0.678892E+09 )
 Step     6: energy -0.767774E+01  ( 0.143880E+05 )
 Step     7: energy -0.818570E+01  ( 0.100364E+05 )
 Step     8: energy -0.938401E+01  ( 0.172508E+04 )
 Step     9: energy -0.109799E+02  ( 0.105718E+04 )
 Step    10: energy -0.116244E+02  ( 0.580821E+03 )
 Step    11: energy -0.116969E+02  ( 0.145676E+04 )
 Step    12: energy -0.118569E+02  ( 0.856882E+03 )
 Step    13: energy -0.119576E+02  ( 0.521381E+03 )
 Step    14: energy -0.120897E+02  ( 0.398602E+03 )
 Step    15: energy -0.121476E+02  ( 0.222201E+03 )
 Step    16: energy -0.121804E+02  ( 0.242613E+03 )
 Step    17: energy -0.121975E+02  ( 0.237765E+03 )
 Step    18: energy -0.122092E+02  ( 0.217435E+03 )
 Step    19: energy -0.122177E+02  ( 0.198576E+03 )
 Step    20: energy -0.122259E+02  ( 0.181726E+03 )
 Step    21: energy -0.122298E+02  ( 0.178796E+03 )
 Step    22: energy -0.122318E+02  ( 0.178968E+03 )
 Step    23: energy -0.122352E+02  ( 0.181503E+03 )
 Step    24: energy -0.122407E+02  ( 0.184272E+03 )
 Step    25: energy -0.122499E+02  ( 0.186336E+03 )
 Step    26: energy -0.122634E+02  ( 0.183760E+03 )
 Step    27: energy -0.122776E+02  ( 0.171223E+03 )
 Step    28: energy -0.122929E+02  ( 0.148718E+03 )
 Step    29: energy -0.123122E+02  ( 0.119336E+03 )
 Step    30: energy -0.123348E+02  ( 0.847677E+02 )
 Step    31: energy -0.123572E+02  ( 0.434069E+02 )
 Step    32: energy -0.123733E+02  ( 0.176053E+02 )
 Step    33: energy -0.123858E+02  ( 0.941361E+01 )
 Step    34: energy -0.123956E+02  ( 0.739923E+01 )
 Step    35: energy -0.124014E+02  ( 0.513793E+01 )
 Step    36: energy -0.124049E+02  ( 0.385630E+01 )
 Step    37: energy -0.124083E+02  ( 0.416421E+01 )
 Step    38: energy -0.124122E+02  ( 0.530410E+01 )
 Step    39: energy -0.124159E+02  ( 0.493920E+01 )
 Step    40: energy -0.124184E+02  ( 0.306992E+01 )
 Step    41: energy -0.124201E+02  ( 0.174747E+01 )
 Step    42: energy -0.124216E+02  ( 0.110719E+01 )
 Step    43: energy -0.124231E+02  ( 0.793785E+00 )
 Step    44: energy -0.124242E+02  ( 0.585080E+00 )
 Step    45: energy -0.124249E+02  ( 0.454481E+00 )
 Step    46: energy -0.124257E+02  ( 0.491193E+00 )
 Step    47: energy -0.124268E+02  ( 0.498392E+00 )
 Step    48: energy -0.124277E+02  ( 0.297774E+00 )
 Step    49: energy -0.124281E+02  ( 0.987143E-01 )
 Step    50: energy -0.124282E+02  ( 0.562089E-01 )
 Step    51: energy -0.124283E+02  ( 0.472681E-01 )
 Step    52: energy -0.124283E+02  ( 0.360471E-01 )
 Step    53: energy -0.124284E+02  ( 0.149691E-01 )
 Step    54: energy -0.124285E+02  ( 0.376934E-02 )
 Step    55: energy -0.124285E+02  ( 0.243872E-02 )
 Step    56: energy -0.124285E+02  ( 0.253572E-02 )
 Step    57: energy -0.124285E+02  ( 0.176076E-02 )
 Step    58: energy -0.124285E+02  ( 0.952400E-03 )
 Step    59: energy -0.124285E+02  ( 0.889573E-03 )
 Step    60: energy -0.124285E+02  ( 0.747233E-03 )
 Step    61: energy -0.124285E+02  ( 0.417036E-03 )
 Step    62: energy -0.124285E+02  ( 0.307630E-03 )
 Step    63: energy -0.124285E+02  ( 0.474529E-03 )

ChangeLog
2006-03-22  Jan H. Meinke <j.meinke@fz-juelich.de>
	* universe.py: Implemented function sheets().
	* protein.py: Implemented more functions in Atom
	* protein.py: Implemented search for strands
2005-10-10  Jan H. Meinke <j.meinke@fz-juelich.de>
        * redvar.f: Read global parameters from var files.  
        * Adjusted calls to ouvar to new interface. I now pass a file name instead of a unit number.  
        * outvar.f: Added molecule number.
        * outvar.f: Added writing of all molecules into one variable file.
        
2005-10-06  Jan H. Meinke <j.meinke@fz-juelich.de>
        * outpdb.f: Added writing of multi molecule pdb files. 
2005-10-05  Jan Meinke  <j.meinke@fz-juelich.de>
	* Removed some extra files.

2005-09-26  Jan H. Meinke  <j.meinke@fz-juelich.de>
	* Committed intermolecular interaction routine.

2005-09-21: Jan H. Meinke <j.meinke@fz-juelich.de>
        * enyshe.f: If nml == 0 calculate energy over all atoms. 

2005-09-19: Jan H. Meinke <j.meinke@fz-juelich.de>
        * Added eninteract.f with function eninteract to project. Calculates interaction energy between molecules.
        * enysol.f: Added if nmol == 0 then calculate energy over all atoms.
        * enysol.f: nresi is total number of residues not just number of residues in first molecule.

2005-09-05: Jan H. Meinke <j.meinke@fz-juelich.de>
	* Fixed calculation of number of free variables for the third and higher
	  molecules.
	* Fixed loop range in rgyr() to take multiple molecules into account.


minim.f
c **************************************************************
c
c This file contains the subroutines: minim,move
c
c Copyright 2003-2005  Frank Eisenmenger, U.H.E. Hansmann,
c                      Shura Hayryan, Chin-Ku 
c Copyright 2007       Frank Eisenmenger, U.H.E. Hansmann,
c                      Jan H. Meinke, Sandipan Mohanty
c
c **************************************************************


      subroutine minim(imin, maxit, eps)

c ......................................................................
c PURPOSE: Use minimizers
c
c          imin = 1:  use Quasi-Newton
c          imin = 2:  use Conjugated Gradients
c
c          @param maxit maximum number of iterations 
c          @param eps acceptance criterium
cInstitute
c CALLS: difang,energy,gradient, mincjg,minqsn, nursvr
c ......................................................................

      include 'INCL.H'
cf2py intent(in) imin
cf2py intent(in) maxit
cf2py intent(in) eps
      parameter (msvmx=mxvr*(mxvr+5)/(2*(2*mxvr+1)),   
     #            msv  = 50 )                             

      dimension w(mxvr*(mxvr+13)/2)

      dimension vlvrn(mxvr),vlvro(mxvr),gdvr(mxvr),scl(mxvr)

c --------------------------- new
      dimension gbpro(6,mxml)

      mxop=maxit
      acc=eps   

      if (msv.gt.msvmx) then
        write (*,'(a,i5)') ' minreg> parameter MSV > ',msvmx
        stop
      endif

c ----------------------- energy & gradient

      call gradient()

      write(*,'(/,a,/)') ' Energy BEFORE minimization:' 

      if (ireg.eq.0) then

        write (*,'(a,e12.5,/,3(a,e11.4),/,2(a,e11.4),/)') ' Total: ',
     #    eysm,
     #    '   Coulomb: ',eyel,' Lennard-Jones: ',eyvw,' HB: ',eyhb,
     #    '   Variables: ',eyvr,'  Solvatation: ',eysl

       else


        write (*,'(a,e12.5,/,3(a,e11.4),/,3(a,e11.4),/)') ' Total: ',
     #    wtey*eysm + wtrg*eyrg,
     #    '   Coulomb: ',eyel,' Lennard-Jones: ',eyvw,' HB: ',eyhb,
     #    '   Variables: ',eyvr,'  Solvatation: ',eysl,
     #    ' Regularization: ',eyrg

       endif

c --------------------------------------- variables

      ntlvr=ivrml1(ntlml)+nvrml(ntlml)-1
      nv=0

      gdey2=0.d0
      gdrg2=0.d0

      do i=1,ntlvr
        if (.not.fxvr(i)) then

          nv=nv+1

          vlvrn(nv) = vlvr(i)

          scl(nv) = 1.d0 ! scale grad_i

          if (ireg.eq.0) then
            gdvr(nv) = gdeyvr(i)
          else
            gdvr(nv) = wtey*gdeyvr(i)+wtrg*gdeyrg(i)
            gdrg2 = gdrg2+gdeyrg(i)**2
          endif

          gdey2=gdey2+gdeyvr(i)**2

        endif

        vlvro(i)=vlvr(i)
      enddo

      n=nv

      if (ireg.eq.0) then

        esm=eysm

      else

        ngbvr=0

        do i=1,ntlml
          do j=1,6

            n=n+1
            ngbvr=ngbvr+1

            vlvrn(n) = gbpr(j,i)
            scl(n) = 1.d0 ! Scale

            gbpro(j,i) = gbpr(j,i)   ! save

            gdvr(n) = wtrg*gdeygb(ngbvr)

            gdrg2=gdrg2+gdeygb(ngbvr)**2
          enddo
        enddo

c        if (abs(wtrg-1.d0).gt.1.d-4.and.abs(wtey-1.d0).gt.1.d-4) then
c          gdey2 = max(acc,gdey2)
c          gdrg2 = max(acc,gdrg2)
c          wtrg = wtrg * sqrt(gdey2/gdrg2)
c          write(*,*)  ' -->    Wt_energy = ',wtey,'  Wt_regul. = ',wtrg
c          write(*,*)  '  '
c        endif

        esm=wtey*eysm+wtrg*eyrg

      endif


      if (imin.eq.1) then ! Quasi-Newton

        n1=1+(n*(n+1))/2
        n2=n1+n
        n3=n2+n
        n4=n3+n
        n5=n4+n
        n6=n5+n

        call minqsn(n,mxvr,vlvrn,esm,gdvr,scl,acc,w,w(n1),w(n2),
     #              w(n3),w(n4),w(n5),w(n6),mxop,nop)

      elseif (imin.eq.2) then ! Conjugated Gradients

        n1=1+n
        n2=n1+n
        n3=n2+n
        n4=n3+n
        n5=n4+n

        call mincjg(n,mxvr,vlvrn,esm,gdvr,acc,w,w(n1),w(n2),  ! no 'scl'
     #              w(n3),w(n4),w(n5),mxop,nop)

      endif


      if (nop.lt.mxop) then
        write (*,'(a)') ' ---- CONVERGENCE ----'
      else
        write (*,'(a)')  '---- STEP LIMIT ----'
      endif


      write (*,'(/,2a,/)') ' Final energies ',
     # '__________________________________________________'

      eysm = energy()

      if (ireg.eq.0) then

        write (*,'(a,e12.5,/,3(a,e11.4),/,2(a,e11.4))') ' Total: ',eysm,
     #  '   Coulomb: ',eyel,' Lennard-Jones: ',eyvw,' HB: ',eyhb,
     #  '   Variables: ',eyvr,'  Solvatation: ',eysl

      else

        write (*,'(a,e12.5,/,3(a,e11.4),/,3(a,e11.4))') ' Total: ',
     #      wtey*eysm + wtrg*eyrg,
     #  '   Coulomb: ',eyel,' Lennard-Jones: ',eyvw,' HB: ',eyhb,
     #  '   Variables: ',eyvr,'  Solvatation: ',eysl,
     #  ' Regularization: ',eyrg

      endif

      write (*,'(/,a,/)') ' Variables _________________'

      nv = 0
      do i=1,ntlvr  !! nvr
        if (.not.fxvr(i)) then

          nv=nv+1
          vr=vlvrn(nv)
          if (abs(vr).gt.pi) vr=vr-sign(pi2,vr)

          write (*,'(1x,a,1x,i4,f8.1,a,f5.1,a)') nmvr(i),nursvr(i),
     #        vr*crd,'    (',abs(difang(vr,vlvro(i)))*crd,')'

          vlvr(i) = vr
        endif
      enddo


      if (ireg.ne.0) then

        write (*,'(/,a,/)') ' Global Variables ___________'

        do i=1,ntlml
          write(*,*) ' Molecule #',i,' old          new'
          do j=1,3
            write(*,*) gbpro(j,i),' ',gbpr(j,i)
          enddo
          do j=4,6
            write(*,*) gbpro(j,i)*crd,' ',gbpr(j,i)*crd
          enddo
        enddo

      endif

      write (*,'(/,2a)') ' Gradient ',
     # '______________________________________________________________'

      write (*,'(8(1x,f8.3))') (gdvr(i),i=1,nv)

      if (ireg.ne.0) then

        write (*,*) ' -------------- global variables ------------'
        write (*,'(6(1x,f8.3))') (gdvr(i+nv),i=1,ngbvr)

      endif

      return
      end
c ********************************************
      subroutine move(nop,nvr1,esm,vlvrn,gdvr)
c
c CALLS: gradient
c
      include 'INCL.H'

      dimension vlvrn(mxvr),gdvr(mxvr)


c ------------------------ compile & new variables

      ntlvr=ivrml1(ntlml)+nvrml(ntlml)-1
      n=0

      do i=1,ntlvr
        if (.not.fxvr(i)) then
          n=n+1
          vlvr(i)=vlvrn(n)
        endif
      enddo

      if (ireg.ne.0) then

        ii=0
        do i=1,ntlml
          do j=1,6   ! global vars.
            ii=ii+1
            gbpr(j,i)=vlvrn(ii+n)
          enddo
        enddo

      endif

c -------------------------- new minimz. gradient

      call gradient()

      gdsmey=0.d0
      gdsmrg=0.d0

      n=0

      do i=1,ntlvr
        if (.not.fxvr(i)) then
          n=n+1

          if (ireg.eq.0) then
            gdvr(n) = gdeyvr(i)
          else
            gdvr(n) = wtey*gdeyvr(i) + wtrg*gdeyrg(i)
            gdsmrg = gdsmrg + gdeyrg(i)**2
          endif

          gdsmey = gdsmey + gdeyvr(i)**2

        endif
      enddo


      if (ireg.eq.0) then

        esm=eysm

        write (*,'(a,i5,a,2(e13.6,a))') ' Step ',nop,': energy ',esm
     #                                 ,'  (',gdsmey,' )'

      else

        ii=0
        do i=1,ntlml  ! global vars.
          do j=1,6
            n=n+1
            ii=ii+1

            gdvr(n) = wtrg*gdeygb(ii)
            gdsmrg = gdsmrg+gdeygb(ii)**2

          enddo
        enddo

        esm=wtey*eysm+wtrg*eyrg

        write (*,'(a,i5,a,3(e13.6,a))') ' Step ',nop,': energy ',esm
     #                                 ,'  (',gdsmey,',',gdsmrg,' )'

      endif

      return
      end

